_posts/2019-04-15-Keep-your-RxJS-streams-from-bursting-their-banks.md

```diff
@@ -0,0 +1,612 @@
+).subscribe(events => {
+  this.events = events;
+});
+
+onDestroy() {
+  subscription1.unsubscribe();
+}
+```
+
+This way, only 1 subscription will exist, canceling it will automatically cancel any inner streams.
+And when `param$` would emit a new value, the `switchMap` operator will automatically cancel its inner stream and create a new one.
+
+> RULE: NEVER SUBSCRIBE WITHIN A SUBSCRIPTION
+This also applies for calling a function inside a subscription when that function has a subscription.
+
+## Sharing is caring
+
+<p style="text-align: center;">
+  <img alt="Cat with big bowl of food: What do you mean, I have to share?" src="{{ '/img/2019-04-15-Keep-your-RxJS-streams-from-bursting-their-banks/2y7xld.jpg' | prepend: site.baseurl }}">
+</p>
+
+When using a stream on multiple locations, remember that there is a difference between a cold and a hot Observable.
+In short: a cold Observable will restart its stream for each subscription, while a hot Observable will reuse an existing stream when a subscription is added.
+Think of a cold Observable as a HTTP request, which fires each time again, and a hot Observable as a keyPress stream which emits the same event no matter how many subscriptions exist on the stream.
+If you need more details, you can read this article by Ben Lesh [Hot vs Cold Observables](https://medium.com/@benlesh/hot-vs-cold-observables-f8094ed53339).
+
+Sometimes you'll need to react on a cold Observable in multiple separate streams.
+For example, you need to display the response, but you also need to initiate a new stream to get other data.
+So you'll create two subscriptions and use a **mergeMap**, because subscriptions inside subscriptions are not allowed:
+
+```typescript
+stream1 = httpGet('userCarMake');
+stream2 = stream1.pipe(mergeMap(make => httpGet('userCarMake', make)));
+stream1.subscribe(make => console.log('The car make is', make));
+stream2.subscribe(models => console.log('The models for this make are', models.join()));
+```
+
+But in your developer tools' network tab, you notice that **userCarMake** has been requested twice.
+The answer to why should be obvious by now: there are two subscriptions.
+To solve this whith minimal changes, make the source stream a hot Observable using [**shareReplay**](https://www.learnrxjs.io/operators/multicasting/sharereplay.html).
+
+```typescript
+stream1 = httpGet('userCarMake').pipe(shareReplay(1));
+stream2 = stream1.pipe(mergeMap(make => httpGet('userCarMake', make)));
+stream1.subscribe(make => console.log('The car make is', make));
+stream2.subscribe(models => console.log('The models for this make are', models.join()));
+```
+
+> RULE: MAKE YOUR OBSERVABLES HOT WHEN NECESSARY
+
+Though be careful when using this operator.
+If the source doesn't complete by itself, it will keep emitting values.
+To prevent this behaviour, use an options object in the shareReplay operator and set the required property **refCount** to true:
```

Comment:
```suggestion
To prevent this behaviour, use an options object in the `shareReplay` operator and set the required property **refCount** to `true`:
```